File: app/src/main/java/wangdaye/com/geometricweather/common/ui/widgets/insets/FitSystemBarViewPager.java
package wangdaye.com.geometricweather.common.ui.widgets.insets;

import android.content.Context;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowInsets;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import java.util.List;

import wangdaye.com.geometricweather.common.basic.insets.FitBothSideBarHelper;
import wangdaye.com.geometricweather.common.basic.insets.FitBothSideBarView;

public class FitSystemBarViewPager extends ViewPager
        implements FitBothSideBarView {

    final FitBothSideBarHelper helper;

    public static class FitBottomSystemBarPagerAdapter extends PagerAdapter {

        private final FitSystemBarViewPager mPager;
        private final List<View> mViewList;
        public List<String> mTitleList;

        public FitBottomSystemBarPagerAdapter(FitSystemBarViewPager pager,
                                              List<View> viewList, List<String> titleList) {
            mPager = pager;
            mViewList = viewList;
            mTitleList = titleList;
        }

        @Override
        public int getCount() {
            return mViewList.size();
        }

        @Override
        public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
            return view == object;
        }

        @NonNull
        @Override
        public Object instantiateItem(@NonNull ViewGroup container, int position) {
            setWindowInsetsForViewTree(mViewList.get(position), mPager.helper.getWindowInsets());
            container.addView(mViewList.get(position));
            return mViewList.get(position);
        }

        @Override
        public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
            container.removeView(mViewList.get(position));
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mTitleList.get(position);
        }

        private void setWindowInsetsForViewTree(View view, Rect insets) {
            setWindowInsets(view, insets);
            if (view instanceof ViewGroup) {
                int count = ((ViewGroup) view).getChildCount();
                for (int i = 0; i < count; i++) {
                    setWindowInsetsForViewTree(((ViewGroup) view).getChildAt(i), insets);
                }
            }
        }

        private void setWindowInsets(View view, Rect insets) {
            if (view instanceof FitSystemBarNestedScrollView) {
                ((FitSystemBarNestedScrollView) view).fitSystemWindows(insets);
            } else if (view instanceof FitSystemBarRecyclerView) {
                ((FitSystemBarRecyclerView) view).fitSystemWindows(insets);
            }
        }
    }

    public FitSystemBarViewPager(@NonNull Context context) {
        this(context, null);
    }

    public FitSystemBarViewPager(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        helper = new FitBothSideBarHelper(this, SIDE_TOP);
    }

    @Override
    public WindowInsets onApplyWindowInsets(WindowInsets insets) {
        return helper.onApplyWindowInsets(insets);
    }

    @Override
    public void addFitSide(@FitSide int side) {
        helper.addFitSide(side);
    }

    @Override
    public void removeFitSide(@FitSide int side) {
        helper.removeFitSide(side);
    }

    @Override
    public void setFitSystemBarEnabled(boolean top, boolean bottom) {
        helper.setFitSystemBarEnabled(top, bottom);
    }

    @Override
    public int getTopWindowInset() {
        return helper.top();
    }

    @Override
    public int getBottomWindowInset() {
        return helper.bottom();
    }
}

File: app/src/main/java/org/breezyweather/common/ui/widgets/insets/FitSystemBarViewPager.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.common.ui.widgets.insets

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager

class FitSystemBarViewPager @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null
) : ViewPager(context, attrs) {

    class FitBottomSystemBarPagerAdapter(
        private val mViewList: List<View>,
        private val mTitleList: List<String>
    ) : PagerAdapter() {
        override fun getCount(): Int {
            return mViewList.size
        }

        override fun isViewFromObject(view: View, `object`: Any): Boolean {
            return view === `object`
        }

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            container.addView(mViewList[position])
            return mViewList[position]
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            container.removeView(mViewList[position])
        }

        override fun getPageTitle(position: Int): CharSequence {
            return mTitleList[position]
        }
    }
}
